module/zfs/vdev_mirror.c

```diff
@@ -625,11 +677,25 @@ vdev_mirror_io_start(zio_t *zio)
 
 	while (children--) {
 		mc = &mm->mm_child[c];
+		c++;
+
+		/*
+		 * When sequentially resilvering only issue write repair
+		 * IOs to the vdev which is being rebuilt since performance
+		 * is limited by the slowest child.  This is an issue for
+		 * faster replacement devices such as distributed spares.
```

Comment:
That's one case.  Another is when you kick in a distributed spare preemptively because you suspect the disk is going to fail soon or your planning to replace it for some other reason.  For example, you're systematically replacing all the disks with larger ones and don't want to give up any redundancy during the process.  The common thread here is we don't want that rebuild to the distributed spare to be slowed down because there are multiple online disks in the mirror.  Particularly, since in the two scenarios I mentioned the non-distributed spare half of the mirror already contains all of the data (since it never went offline).  The sequential resilver always starts a scrub when its done so this second pass will copy any missing data to the other mirror children.

Reply:
When you do the `zpool replace`, is that reading from the other (D+P-1) disks in the group, doing the RAIDZ reconstruction math, and then writing to the `replacing` vdev, and then we don't bother writing to the presumed-good side of the `replacing`?

Or does `zpool replace` read from the presumed-good side of the `replacing`, and write to the known-bad (new) side of the `replacing`?

In either case, it seems like the performance of the operation would be limited by the reads.

Reply:
The answer is it depends.  When you have a `replacing` or `spare` vdev with multiple online vdev's it happens at the mirror level as long as the DTLs indicate it's partially replicated.  It reads from the good child and issues a repair IO to any children where its missing.  If none of the `replacing` or `spare` mirror vdevs have the data then it's reconstructed at the RAIDZ/DRAID level from parity and the repair IO is issued there.

> In either case, it seems like the performance of the operation would be limited by the reads.

You are limited by the read performance.  However, there's also pretty clearly a significant penalty when writing to all of the mirror children.  When the group width (D+P) is approximatly the same as the number of dRAID children (C) this really doesn't matter.  However, when (C / (D+P)) is significantly larger than 1 is becomes increasingly important.  When testing with a draid2:8d:90c:1s pool without this change I see approximate 200MB/s when sequentially resilvering.  With this change (i.e. only issuing repair IO to the distributed spare) it's closer to 800MB/s.